src/seq/mod.rs

```diff
@@ -188,20 +188,57 @@ pub trait IteratorRandom: Iterator + Sized {
     fn choose<R>(mut self, rng: &mut R) -> Option<Self::Item>
         where R: Rng + ?Sized
     {
-        if let Some(elem) = self.next() {
-            let mut result = elem;
-            
-            // Continue until the iterator is exhausted
-            for (i, elem) in self.enumerate() {
-                let denom = (i + 2) as f64; // accurate to 2^53 elements
+        let (mut lower, mut upper) = self.size_hint();
+        let mut consumed = 0;
+        let mut result = None;
+
+        if upper == Some(lower) {
+            // Remove this once we can specialize on ExactSizeIterator
+            return if lower == 0 { None } else { self.nth(rng.gen_range(0, lower)) };
+        } else if lower <= 1 {
```

Comment:
My goal was to ensure that the new code should reduce down to the old code once the compiler does constant folding of an "unhinted" iterator. And similarly that doing constant folding of a perfectly hinted iterator would reduce to SliceRandom::choose. I.e. the first two categories discussed in the initial comment should reduce to optimal code once constant folding is done.

That's why both of these paths are there. Both could be removed without loosing any correctness, and in both cases we just end up adding a few branches to the runtime code.

(The difference for "perfectly hinted" is slightly bigger for iterators of size exactly 1, but that's a rare use case anyway).

We definitely could remove these paths. I don't have a strong opinion either way. The performance difference will be small and will mainly depend on how slow other operations are, i.e. how many items the "unhinted" iterator has, and how fast RngCore implementation is used.

Reply:
Aha, makes some sense, though I still don't really understand the reason to use `lower <= 1` here. You could simply drop the condition (equivalent of `true`) and would get the same behaviour on unhinted iterators.

My gut feeling is that the cases worth bothering with are (a) perfectly hinted iterators, (b) iterators where some lower bound is known, and (c) iterators with no hinting. So you could potentially insert a `lower > 0` case (instead of the loop case) and use this code for the `lower == 0` case — but I think only benchmarks can tell which one is best.

Are chunked iterators common? It's not a pattern I remember seeing with iterators, but I guess it may be relevant to buffered input.

Reply:
> Aha, makes some sense, though I still don't really understand the reason to use lower <= 1 here

When `lower > 1` it is better to use `gen_range` to pick one of the first `lower` items, than to unconditionally grab the first item and then iterate.

> Are chunked iterators common?

No idea.

As mentioned in other comments, the idea is that for all other cases, the code will reduce to the optimal approach. So the cost to support chunked iterators is only one of source complexity, neither one of runtime cost, nor of binary size.

Reply:
Removing this path improves benchmarks slightly. The only functional difference from the second path within the loop is that the `gen_bool` bit is skipped.

The `Bernoulli` code actually has special handling for the `p==1.0` case anyway and doesn't touch the RNG.

Simpler code and faster benchmarks implies we're probably better off without this code?

Reply:
I don't have a strong preference either way. Generally it's a good idea to be careful about optimizing too much based on just microbenchmarks, since that often doesn't reflect real-world performance.

In theory this code should help "unhinted" iterators, though as you point out by very little since it mainly saves a few branches. In neither case will we call into the RNG to generate numbers. So yeah, probably worth removing since the win seems quite small.

I can see two reasons why removing this code would speed up anything:
* We might end up inlining things better since the function is smaller.
* The "window hinted" test might get faster since we do one branch less. However this benchmark seems less important to me than the "unhinted" and the fully hinted ones.

Which benchmarks specifically got faster with this removed?

Reply:
The unhinted one:
```
test seq_iter_choose_from_1000               ... bench:       3,756 ns/iter (+/- 108) = 2129 MB/s
test seq_iter_unhinted_choose_from_1000      ... bench:       5,311 ns/iter (+/- 98)
test seq_iter_window_hinted_choose_from_1000 ... bench:       1,789 ns/iter (+/- 44)
# after removing this branch:
test seq_iter_choose_from_1000               ... bench:       3,746 ns/iter (+/- 60) = 2135 MB/s
test seq_iter_unhinted_choose_from_1000      ... bench:       5,007 ns/iter (+/- 106)
test seq_iter_window_hinted_choose_from_1000 ... bench:       1,769 ns/iter (+/- 42)
```